frontend/src/components/Scratch/Scratch.tsx

```diff
@@ -169,6 +169,15 @@ export default function Scratch({
         incrementValueVersion()
     }, [scratch.slug, scratch.last_updated])
 
+    const cmExtensions = [...CODEMIRROR_EXTENSIONS, sourceCompareExtension]
+    const cmExtensionsRef = useRef(cmExtensions)
+    useEffect(() => {
+        if (useVim == true) {
+            cmExtensionsRef.current = [...cmExtensionsRef.current, vim()]
+        }
+    })
```

Comment:
This setup with useEffect and refs confuses me, why is it not just `let cmExtensions = [...CODE_MIRROR_EXTENSIONS, sourceCompareExtension; if (useVim) cmExtensions.push(vim())`?

In particular it looks like this useEffect callback will fire any time this component re-renders, and continuously append `vim()` entries to the array, never removing them. Also, this will happen with the wrong timing: after render, rather than immediately.